components/supervisor/pkg/gitpod/gitpod-config.go

```diff
@@ -0,0 +1,244 @@
+func (service *ConfigService) Observe(ctx context.Context) (<-chan *GitpodConfig, <-chan error) {
+	listener := configListener{
+		configs: make(chan *GitpodConfig),
+		errors:  make(chan error),
+	}
+
+	go func() {
+		defer close(listener.configs)
+		defer close(listener.errors)
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-service.locationReady:
+		}
+
+		err := service.start()
+		if err != nil {
+			// failed to start
+			listener.errors <- err
+			return
+		}
+		listener.configs <- service.config
+
+		service.mutex.Lock()
+		service.listeners[listener] = struct{}{}
+		service.mutex.Unlock()
+
+		defer func() {
+			service.mutex.Lock()
+			defer service.mutex.Unlock()
+			delete(service.listeners, listener)
+			if len(service.listeners) == 0 && service.stop != nil {
+				service.stop()
+				service.stop = nil
+			}
+		}()
+
+		select {
+		case <-ctx.Done():
+			return
+		}
+	}()
+	return listener.configs, listener.errors
+}
+
+func (service *ConfigService) start() error {
+	context := service.tryStart()
+	if context == nil {
+		// alread running
+		return nil
+	}
+	_, err := os.Stat(service.location)
+	if service.tryPolling(context, err) {
+		// inotify cannot watch inexistent file, let's poll
+		return nil
+	}
+	return service.watch(context)
+}
+
+func (service *ConfigService) tryStart() context.Context {
+	service.mutex.Lock()
+	defer service.mutex.Unlock()
+
+	if service.stop != nil {
+		return nil
+	}
+
+	log.WithField("location", service.location).Info("Starting watching...")
+	context, stop := context.WithCancel(context.Background())
+	service.stop = stop
+	return context
+}
+
+func (service *ConfigService) watch(context context.Context) error {
+	watcher, startErr := fsnotify.NewWatcher()
+	defer func() {
+		if startErr != nil {
+			log.WithField("location", service.location).WithError(startErr).Error("Failed to start watching...")
+		} else {
+			log.WithField("location", service.location).Info("Started watching")
+		}
+	}()
+	if startErr != nil {
+		return startErr
+	}
+
+	startErr = watcher.Add(service.location)
+	if startErr != nil {
+		watcher.Close()
+		return startErr
+	}
+
+	go func() {
+		defer log.WithField("location", service.location).Info("Stopped watching")
+		defer watcher.Close()
+
+		polling := make(chan struct{}, 1)
+		service.scheduleUpdateConfig(context, polling)
+		for {
+			select {
+			case <-polling:
+				return
+			case <-context.Done():
+				return
+			case err := <-watcher.Errors:
+				service.dispatchError(err)
+			case <-watcher.Events:
+				service.scheduleUpdateConfig(context, polling)
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (service *ConfigService) scheduleUpdateConfig(context context.Context, polling chan<- struct{}) {
+	service.mutex.Lock()
+	defer service.mutex.Unlock()
+	if service.timer != nil {
+		service.timer.Stop()
+	}
+	service.timer = time.AfterFunc(100*time.Millisecond, func() {
+		err := service.updateConfig()
+		if service.tryPolling(context, err) {
```

Comment:
This is a good example of why `tryPolling` is odd. The following would be more "Go" like, i.e. closer to other Go code, but also would keep all "concerns" in one place.


```
if os.IsNotExist(err) {
    service.poll()
    polling <- struct{}{}
} else if err != nil {
    service.dispatchError(err)
}
```